AOT  jsr292Test test_bindTo_interfaceMethodImplementation_package - Successfully invoked package private implementation of an interface method, StackWalkerTest testClassLoaderName Wrong classloader name expected

https://openj9-jenkins.osuosl.org/job/Test_openjdk17_j9_extended.functional_s390x_linux_aot_Personal/232/
https://openj9-jenkins.osuosl.org/job/Test_openjdk11_j9_extended.functional_s390x_linux_aot_Personal_testList_1/270/
 jsr292Test_1
https://openj9-jenkins.osuosl.org/job/Test_openjdk11_j9_extended.functional_s390x_linux_aot_Personal_testList_0/270
jsr292Test_0
```
=== AOT ITERATION 2 ===

07:27:01  FAILED: test_bindTo_interfaceMethodImplementation_package
07:27:01  java.lang.AssertionError: Successfully invoked package private implementation of an interface method.
07:27:01  	at org.testng.Assert.fail(Assert.java:96)
07:27:01  	at com.ibm.j9.jsr292.InterfaceHandleTest.test_bindTo_interfaceMethodImplementation_package(InterfaceHandleTest.java:102)
```

This is related so I didn't open another issue.

https://openj9-jenkins.osuosl.org/job/Test_openjdk11_j9_extended.functional_s390x_linux_aot_Personal_testList_0/270
StackWalkerTest_0
```
09:06:16  FAILED: testClassLoaderName
09:06:16  java.lang.AssertionError: Wrong classloader name expected [TestMyloaderInstance] but found [null]
09:06:16  	at org.testng.Assert.fail(Assert.java:96)
09:06:16  	at org.testng.Assert.failNotEquals(Assert.java:776)
09:06:16  	at org.testng.Assert.assertEqualsImpl(Assert.java:132)
09:06:16  	at org.testng.Assert.assertEquals(Assert.java:118)
09:06:16  	at org.testng.Assert.assertEquals(Assert.java:453)
09:06:16  	at org.openj9.test.stackWalker.MyRunnable.run(MyRunnable.java:45)
09:06:16  	at org.openj9.test.stackWalker.StackWalkerTest.testClassLoaderName(StackWalkerTest.java:243)
```

Changes from last successful build https://openj9-jenkins.osuosl.org/job/Pipeline-Build-Test-AOT/291/
https://github.com/eclipse-openj9/openj9/compare/bc7affe3fed31...581a248f98c
https://github.com/eclipse-openj9/openj9-omr/compare/3a218f25ea0...770936378ec
https://github.com/ibmruntimes/openj9-openjdk-jdk11/compare/839808e0...720cbbf6053

The `InterfaceHandleTest` failure is caused by incorrect romClass comparison result. The class comparison skipped the `ROM_CLASS_MODIFIERS`, and `compareROMClassForEquality()` returned `TRUE` for 2 classes with different modifiers. 

https://github.com/eclipse-openj9/openj9/blob/dd01de181d2c24d8e8bd3899b6cb7cd1a5533316/runtime/bcutil/ComparingCursor.cpp#L466-L469

```
[2024-06-20T11:27:01.681Z] PASSED: test_bindTo_interfaceMethodImplementation_public
[2024-06-20T11:27:01.681Z] PASSED: testIfDirectHandleTestIsJitted
[2024-06-20T11:27:01.681Z] FAILED: test_bindTo_interfaceMethodImplementation_package
[2024-06-20T11:27:01.681Z] java.lang.AssertionError: Successfully invoked package private implementation of an interface method.
[2024-06-20T11:27:01.681Z] 	at org.testng.Assert.fail(Assert.java:96)
[2024-06-20T11:27:01.681Z] 	at com.ibm.j9.jsr292.InterfaceHandleTest.test_bindTo_interfaceMethodImplementation_package(InterfaceHandleTest.java:102)
```

The class generated by `test_bindTo_interfaceMethodImplementation_public` is stored into SCC and then incorrectly re-used by `test_bindTo_interfaceMethodImplementation_package` via romClass comparison, so the test failed.

--- a/runtime/bcutil/ComparingCursor.cpp
+++ b/runtime/bcutil/ComparingCursor.cpp
@@ -464,10 +464,6 @@ ComparingCursor::shouldCheckForEquality(DataType dataType, U_32 u32Value)
 	case INTERMEDIATE_CLASS_DATA_LENGTH:
 		return false;
 	case ROM_CLASS_MODIFIERS:
-		if (u32Value == (u32Value & _context->romMethodModifiers())) {
-			/*It's ok if the existing rom class has extra debug information and it is not required*/
-			return false;
-		}
 		break;
 	case OPTIONAL_FLAGS:
 		if (u32Value == (u32Value & _context->romClassOptionalFlags())) {

--- a/runtime/vm/exceptiondescribe.c
+++ b/runtime/vm/exceptiondescribe.c
@@ -40,6 +40,7 @@ typedef UDATA (*callback_func_t) (J9VMThread * vmThread, void * userData, UDATA 
 static void printExceptionInThread (J9VMThread* vmThread);
 static UDATA isSubclassOfThreadDeath (J9VMThread *vmThread, j9object_t exception);
 static void printExceptionMessage (J9VMThread* vmThread, j9object_t exception);
+static J9Class* findJ9ClassForROMClass(J9VMThread *vmThread, J9ROMClass *romClass, J9ClassLoader **resultClassLoader);
 
 
 /* assumes VM access */
@@ -219,6 +220,98 @@ printStackTraceEntry(J9VMThread * vmThread, void * voidUserData, UDATA bytecodeO
 	return TRUE;
 }
 
+/**
+* Find the J9Class given a ROMClass.
+* @param[in] vmThread The current VM thread.
+* @param[in] ROMClass The J9ROMClass from which the J9Class is created.
+* @param[in, out] resultClassLoader This is the classLoader that owns the memory segment that the ROMClass was found in.
+*	The romclass memory segment in the shared cache always belongs to vm->systemClassLoader. Set it to the actual class loader which loaded the class in this case.
+*
+* @return The J9class, or NULL on failure.
+*/
+static J9Class*
+findJ9ClassForROMClass(J9VMThread *vmThread, J9ROMClass *romClass, J9ClassLoader **resultClassLoader)
+{
+	J9UTF8 const *utfClassName = J9ROMCLASS_CLASSNAME(romClass);
+	J9JavaVM *vm = vmThread->javaVM;
+	J9Class* ret = NULL;
+	if (j9shr_Query_IsAddressInCache(vm, romClass, romClass->romSize)) {
+		J9ClassLoaderWalkState walkState;
+		J9ClassLoader* classLoader = NULL;
+		BOOLEAN fastMode = J9_ARE_ALL_BITS_SET(vm->extendedRuntimeFlags, J9_EXTENDED_RUNTIME_FAST_CLASS_HASH_TABLE);
+		J9Class* ramClass = NULL;
+		if (!fastMode) {
+			omrthread_monitor_enter(vm->classTableMutex);
+		}
+		/**
+		 * Use hashClassTableAt() instead of peekClassHashTable() because classLoaderBlocksMutex must be acquired after classTableMutex.
+		 * (allClassLoadersStartDo enters classLoaderBlocksMutex and peekClassHashTable() enters classTableMutex)
+		 */
+		/**
+		 * All ROMClasses from the SCC are owned by the bootstrap class loader. To minimize the chances to iterate all class loaders, probe
+		 * the booststrap loader, extensionClassLoader and application loader first to determine if they have the J9Class for the current class.
+		 */
+		ramClass = hashClassTableAt(*resultClassLoader, J9UTF8_DATA(utfClassName), J9UTF8_LENGTH(utfClassName));
+		if ((NULL != ramClass)
+			&& (romClass == ramClass->romClass)
+		) {
+			ret = ramClass;
+			if (!fastMode) {
+				omrthread_monitor_exit(vm->classTableMutex);
+			}
+			goto done;
+		}
+
+		ramClass = hashClassTableAt(vm->extensionClassLoader, J9UTF8_DATA(utfClassName), J9UTF8_LENGTH(utfClassName));
+		if ((NULL != ramClass)
+			&& (romClass == ramClass->romClass)
+		) {
+			ret = ramClass;
+			*resultClassLoader = vm->extensionClassLoader;
+			if (!fastMode) {
+				omrthread_monitor_exit(vm->classTableMutex);
+			}
+			goto done;
+		}
+
+		ramClass = hashClassTableAt(vm->applicationClassLoader, J9UTF8_DATA(utfClassName), J9UTF8_LENGTH(utfClassName));
+		if ((NULL != ramClass)
+			&& (romClass == ramClass->romClass)
+		) {
+			ret = ramClass;
+			*resultClassLoader = vm->applicationClassLoader;
+			if (!fastMode) {
+				omrthread_monitor_exit(vm->classTableMutex);
+			}
+			goto done;
+		}
+
+		classLoader = vm->internalVMFunctions->allClassLoadersStartDo(&walkState, vm, 0);
+		while (NULL != classLoader) {
+			if ((classLoader != *resultClassLoader)
+				&& (classLoader != vm->applicationClassLoader)
+			) {
+				ramClass = hashClassTableAt(classLoader, J9UTF8_DATA(utfClassName), J9UTF8_LENGTH(utfClassName));
+				if ((NULL != ramClass)
+					&& (romClass == ramClass->romClass)
+				) {
+					ret = ramClass;
+					*resultClassLoader = classLoader;
+					break;
+				}
+			}
+			classLoader = vm->internalVMFunctions->allClassLoadersNextDo(&walkState);
+		}
+		vm->internalVMFunctions->allClassLoadersEndDo(&walkState);
+		if (!fastMode) {
+			omrthread_monitor_exit(vm->classTableMutex);
+		}
+	} else {
+		ret = peekClassHashTable(vmThread, *resultClassLoader, J9UTF8_DATA(utfClassName), J9UTF8_LENGTH(utfClassName));
+	}
+done:
+	return ret;
+}
 
 /*
  * Walks the backtrace of an exception instance, invoking a user-supplied callback function for
@@ -352,29 +445,7 @@ inlinedEntry:
 #endif
 					romClass = findROMClassFromPC(vmThread, methodPC, &classLoader);
 					if (NULL != romClass) {
-						J9UTF8 const *utfClassName = J9ROMCLASS_CLASSNAME(romClass);
-
-						ramClass = peekClassHashTable(vmThread, classLoader, J9UTF8_DATA(utfClassName), J9UTF8_LENGTH(utfClassName));
-						if (j9shr_Query_IsAddressInCache(vm, romClass, romClass->romSize)) {
-							if (ramClass == NULL) {
-								/* Probe the application loader to determine if it has the J9Class for the current class.
-								 * This secondary probe is required as all ROMClasses from the SCC appear to be owned
-								 * by the bootstrap classloader.
-								 */
-								ramClass = peekClassHashTable(vmThread, vm->applicationClassLoader, J9UTF8_DATA(utfClassName), J9UTF8_LENGTH(utfClassName));
-							}
-							if (NULL != ramClass) {
-								if (romClass != ramClass->romClass) {
-									/**
-									 * It is possible this romClass is not loaded by bootstrap/app class loader.
-									 * There is another class loader that loads a different class with the same name,
-									 * Do not return the ramClass from bootstrap/app loader in this case.
-									 */
-									ramClass = NULL;
-								}
-							}
-						}
-
+						ramClass = findJ9ClassForROMClass(vmThread, romClass, &classLoader);
 						while (NULL != ramClass) {
 							U_32 i = 0;
 							J9Method *methods = ramClass->ramMethods;

